src/graphics/Window.zig

```diff
@@ -1,25 +1,381 @@
+		if (settings.downloadControllerMappings) {
+			var timestamp: i128 = 0;
+			if (files.openDir(".") catch null) |dir| {
+				var _dir: files.Dir =dir;
+				if (_dir.hasFile("gamecontrollerdb.stamp")) {
+					const stamp = _dir.read(main.globalAllocator, "gamecontrollerdb.stamp") catch unreachable;
+					defer main.globalAllocator.free(stamp);
+					if (std.fmt.parseInt(i128, stamp, 16) catch null) |newTimestamp| {
+						timestamp = newTimestamp;
+					}
+				}
+			}
+			const delta = curTimestamp - timestamp;
+			const deltaDays = @divTrunc(delta, std.time.ns_per_day);
+			needDownload = deltaDays >= 7;
+		}
+
+		if (settings.downloadControllerMappingsWhenUnrecognized) {
+			for (0..c.GLFW_JOYSTICK_LAST) |jsid| {
+				if ((c.glfwJoystickPresent(@intCast(jsid)) != 0) and (c.glfwJoystickIsGamepad(@intCast(jsid)) == 0)) {
+					needDownload = true;
+					break;
+				}
+			}
+		}
+		if (needDownload) {
+			if (@atomicRmw(bool, &self.controllerMappingsDownloaded, std.builtin.AtomicRmwOp.Xchg, false, std.builtin.AtomicOrder.monotonic)) {
+				return; // Controller mappings are already downloading.
+			}
+			ControllerMappingDownloadTask.schedule(curTimestamp, &self.controllerMappingsDownloaded);
+		}
+	}
+	pub fn updateControllerMappings(_: *Gamepad) void {
+		var _envMap = std.process.getEnvMap(main.stackAllocator.allocator) catch null;
+		if (_envMap) |*envMap| {
+			defer envMap.deinit();
+			if (envMap.get("SDL_GAMECONTROLLERCONFIG")) |controller_config_env| {
+				_ = c.glfwUpdateGamepadMappings(@ptrCast(controller_config_env));
+			}
+		}
+		if (files.openDir(".") catch null) |dir| {
+			if (dir.read(main.stackAllocator, "gamecontrollerdb.txt") catch null) |data| {
+				var newData = main.stackAllocator.realloc(data, data.len + 1);
+				defer main.stackAllocator.free(newData);
+				newData[data.len - 1] = 0;
+				_ = c.glfwUpdateGamepadMappings(@ptrCast(newData));
+			}
+		}
+		const selfExeDirPath = std.fs.selfExeDirPathAlloc(main.stackAllocator.allocator) catch |err| {
+			std.log.err("Error getting executable directory{s}", .{@errorName(err)});
+			return;
+		};
+		defer main.stackAllocator.free(selfExeDirPath);
+		const mappings_path = std.fs.path.join(main.stackAllocator.allocator, &.{selfExeDirPath, "gamecontrollerdb.txt"}) catch unreachable;
+		defer main.stackAllocator.free(mappings_path);
+		const data =  main.files.read(main.stackAllocator, mappings_path) catch |err| {
+			if (@TypeOf(err) == std.fs.File.OpenError and err == std.fs.File.OpenError.FileNotFound) {
+				return; // Ignore not finding mappings.
+			}
+			std.log.err("Error opening gamepad mappings file: {s}", .{@errorName(err)});
+			return;
+		};
+		defer main.stackAllocator.free(data);
+		_ = c.glfwUpdateGamepadMappings(@ptrCast(data));
+	}
+
+	pub fn init() *Gamepad {
+		const self: *Gamepad = main.globalAllocator.create(Gamepad);
+		if (!settings.askToDownloadControllerMappings) {
+			self.*.downloadControllerMappings();
```

Comment:
You don't need to do .* when accessing members of a pointer.
This is Zig, not C/C++.